test/common/config/delta_subscription_impl_test.cc

```diff
@@ -77,14 +78,103 @@ TEST_F(DeltaSubscriptionImplTest, ResourceGoneLeadsToBlankInitialVersion) {
+  subscription_->updateResources({"name1", "name2", "name3"});
+  InSequence s;
+  expectSendMessage({"name3"}, {}, Grpc::Status::GrpcStatus::Ok, "");
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the second update
+  subscription_->resume();
+}
+
+// Due to how our implementation provides the required behavior tested in RemoveThenAdd, the
+// add-then-remove case *also* causes the resource to be referred to in the request (as an
+// unsubscribe).
+// Unlike the remove-then-add case, this one really is unnecessary, and ideally we would have
+// the request simply not include any mention of the resource. Oh well.
+// This test is just here to illustrate that this behavior exists, not to enforce that it
+// should be like this. What *is* important: the server must happily and cleanly ignore
+// "unsubscribe from [resource name I have never before referred to]" requests.
+TEST_F(DeltaSubscriptionImplTest, AddThenRemove) {
+  startSubscription({"name1", "name2", "name3"});
+  subscription_->pause(); // Pause because we're testing multiple updates in between request sends.
+  subscription_->updateResources({"name1", "name2", "name3", "name4"});
+  subscription_->updateResources({"name1", "name2", "name3"});
+  InSequence s;
+  expectSendMessage({}, {"name4"}, Grpc::Status::GrpcStatus::Ok, "");
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the second update
+  subscription_->resume();
+}
+
+// add/remove/add == add.
+TEST_F(DeltaSubscriptionImplTest, AddRemoveAdd) {
+  startSubscription({"name1", "name2", "name3"});
+  subscription_->pause();
+  subscription_->updateResources({"name1", "name2", "name3", "name4"});
+  subscription_->updateResources({"name1", "name2", "name3"});
+  subscription_->updateResources({"name1", "name2", "name3", "name4"});
+  InSequence s;
+  expectSendMessage({"name4"}, {}, Grpc::Status::GrpcStatus::Ok, "");
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the second update
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the third update
+  subscription_->resume();
+}
+
+// remove/add/remove == remove.
+TEST_F(DeltaSubscriptionImplTest, RemoveAddRemove) {
+  startSubscription({"name1", "name2", "name3"});
+  subscription_->pause();
+  subscription_->updateResources({"name1", "name2"});
+  subscription_->updateResources({"name1", "name2", "name3"});
+  subscription_->updateResources({"name1", "name2"});
+  InSequence s;
+  expectSendMessage({}, {"name3"}, Grpc::Status::GrpcStatus::Ok, "");
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the second update
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the third update
+  subscription_->resume();
+}
+
+// Starts with 1,2,3. 4 is added/removed/added. In those same updates, 1,2,3 are
+// removed/added/removed. End result should be 4 added and 1,2,3 removed.
```

Comment:
I like these tests, thanks for the great comments.

Reply:
Awesome, thanks, glad to hear it!

Reply:
Awesome, thanks glad to hear it! :)